pyccel/parser/semantic.py

```diff
@@ -3455,53 +3455,84 @@ def _visit_For(self, expr):
+            step = LiteralInteger(1)
+            vars = []
+            a = Iterable(self._visit(body.iterable))
+            if isinstance(a.iterable, PythonRange):
+                var = self.scope.get_expected_name(expr.indices[i])
+                indices.append(var)
                 var   = self._create_variable(var, PythonNativeInt(), start, {})
                 dvar  = self._infer_type(var)
-                stop  = a.stop
-                start = a.start
-                step  = a.step
-
-            elif isinstance(a, (PythonZip, PythonEnumerate)):
-                dvar  = self._infer_type(a.element)
-                class_type = dvar.pop('class_type')
+                vars.append((var, dvar))
+                stop  = a.iterable.stop
+                start = a.iterable.start
+                step  = a.iterable.step
+            elif isinstance(a.iterable, PythonEnumerate):
+                var1 = self.scope.get_expected_name(expr.indices[i][0])
+                var2 = self.scope.get_expected_name(expr.indices[i][1])
+                indices.append(var1)
+                indices.append(var2)
+                dvar2 = self._infer_type(a.iterable.element)
+                class_type = dvar2.pop('class_type')
                 if class_type.rank > 0:
-                    class_type = class_type.switch_rank(class_type.rank-1)
-                    dvar['shape'] = (dvar['shape'])[1:]
+                    class_type = class_type.switch_rank(class_type.rank - 1)
+                    dvar2['shape'] = (dvar2['shape'])[1:]
                 if class_type.rank == 0:
-                    dvar['shape'] = None
-                    dvar['memory_handling'] = 'stack'
-                var  = Variable(class_type, var, **dvar)
-                stop = a.element.shape[0]
-
-            elif isinstance(a, Variable):
-                dvar  = self._infer_type(a)
+                    dvar2['shape'] = None
+                    dvar2['memory_handling'] = 'stack'
+                var2 = Variable(class_type, var2, **dvar2)
+                var1 = Variable(PythonNativeInt(), var1)
+                vars.append((var1, self._infer_type(var1)))
+                vars.append((var2, dvar2))
+                stop = a.iterable.element.shape[0]
+            elif isinstance(a.iterable, PythonZip):
+                for idx, arg in enumerate(a.iterable.args):
+                    var = self.scope.get_expected_name(expr.indices[i][idx])
+                    indices.append(var)
+                    dvar = self._infer_type(arg)
+                    class_type = dvar.pop('class_type')
+                    if class_type.rank > 0:
+                        class_type = class_type.switch_rank(class_type.rank-1)
+                        dvar['shape'] = (dvar['shape'])[1:]
+                    if class_type.rank == 0:
+                        dvar['shape'] = None
+                        dvar['memory_handling'] = 'stack'
+                    var  = Variable(class_type, var, **dvar)
```

Comment:
Not ideal to create variables without using the vriable creation functions. This usually leads to duplicated type checks or forgotten tests (e.g setting target as necessary)